helper/schema/resource.go

```diff
@@ -472,12 +474,34 @@ func (r *Resource) InternalValidate(topSchemaMap schemaMap, writable bool) error
 		}
 	}
 
-	if r.LegacySchema.Version >= r.SchemaVersion {
-		return errors.New("LegacySchema.Version cannot be >= SchemaVersion")
+	// verify state upgraders are consecutive and have registered schema types
+	sort.Slice(r.StateUpgraders, func(i, j int) bool {
```

Comment:
Rather than sorting these here, should we instead require that they already be sorted and generate an error if not? I don't imagine it matters that much _functionally_ if they aren't sorted, but having them out of order is likely to make the declaration in source code hard to read anyway and so in practice I can't think of a good reason to allow it.